clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -117,6 +137,341 @@ class image_mem_wrapper {
+public:
+#ifdef _WIN32
+  interop_mem_wrapper(ID3D11Resource *d3d11_res, unsigned reg_flags)
+      : _d3d11_res(d3d11_res), _reg_flags(reg_flags) {
+
+    assert(uniqueD3D11Res.find(_d3d11_res) == uniqueD3D11Res.end() &&
+        "This D3D11 resource is already registered");
+
+    query_res_info(_d3d11_res);
+
+    _win_nt_handle = create_shared_handle(_d3d11_res);
+
+    sycl::ext::oneapi::experimental::resource_win32_handle 
+      ext_mem_win_handle{_win_nt_handle};
+
+    sycl::ext::oneapi::experimental::external_mem_descriptor
+      <sycl::ext::oneapi::experimental::resource_win32_handle> ext_mem_desc{
+        ext_mem_win_handle,
+        sycl::ext::oneapi::experimental::external_mem_handle_type::
+            win32_nt_dx12_resource,
+        _res_size_bytes};
+
+    _interop_mem_handle_ptr =
+      new sycl::ext::oneapi::experimental::interop_mem_handle;
+    *_interop_mem_handle_ptr =
+      sycl::ext::oneapi::experimental::import_external_memory(
+        ext_mem_desc, get_default_queue());
+
+    uniqueD3D11Res.insert(_d3d11_res);
+  }
+
+  ~interop_mem_wrapper() {
+    assert(_interop_mem_handle_ptr && "Resource is not registered!");
+    
+    CloseHandle(_win_nt_handle);
+
+    sycl::ext::oneapi::experimental::release_external_memory(
+        get_interop_mem_handle(), get_default_queue());
+    
+    uniqueD3D11Res.erase(_d3d11_res);
+  }
+#endif
+
+  interop_mem_wrapper(const interop_mem_wrapper &) = delete;
+
+  interop_mem_wrapper &operator=(const interop_mem_wrapper &) = delete;
+
+  const sycl::ext::oneapi::experimental::interop_mem_handle
+  get_interop_mem_handle() const noexcept {
+    return *_interop_mem_handle_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_mem_handle
+  get_img_mem_handle() const noexcept {
+    return *_img_mem_handle_ptr;
+  }
+
+  void *
+  get_buf_ptr() const noexcept {
+    return _buf_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_descriptor
+  get_img_desc() const noexcept {
+    return _img_desc;
+  }
+
+  size_t get_res_size_bytes() {
+    return _res_size_bytes;
+  }
+
+  bool is_buffer_res() const {
+    return _res_is_buffer;
+  }
+
+  void map_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      assert(!_buf_ptr && "Resource is already mapped!");
+
+      // NOTE: 'map_external_linear_memory' API is not yet merged by Codeplay
+      /*
+      _buf_ptr =
+        sycl::ext::oneapi::experimental::map_external_linear_memory(
+          get_interop_mem_handle(), _img_desc, q);
+      */
+      throw std::runtime_error("Mapping a buffer resoure is not yet supported!");
+    }
+    else {
+      assert(!_img_mem_handle_ptr && "Resource is already mapped!");
+
+      _img_mem_handle_ptr =
+        new sycl::ext::oneapi::experimental::image_mem_handle;
+      *_img_mem_handle_ptr =
+        sycl::ext::oneapi::experimental::map_external_image_memory(
+          get_interop_mem_handle(), _img_desc, q);
+      
+      _img_mem_wrapper_ptr = new image_mem_wrapper(this);
```

Comment:
`_img_mem_handle_ptr` doesn't need to store in the class. Using `_img_mem_wrapper_ptr` to manage it.